source/simulator/assembly.cc

```diff
@@ -1227,6 +1227,206 @@ namespace aspect
+                                       != dealii::identity_tensor<dim>());
+              for (unsigned int i = 0; i < dofs_per_cell; ++i)
+                for (unsigned int j = 0; j < dofs_per_cell; ++j)
+                  if (fe.system_to_component_index(i).first ==
+                      fe.system_to_component_index(j).first)
+                    data.local_matrix(i, j) += ((
+                                                  use_tensor ?
+                                                  eta * (scratch.grads_phi_u[i]
+                                                         * stress_strain_director
+                                                         * scratch.grads_phi_u[j]) :
+                                                  eta * (scratch.grads_phi_u[i]
+                                                         * scratch.grads_phi_u[j]))
+                                                + (1. / eta) * pressure_scaling
+                                                * pressure_scaling
+                                                * (scratch.phi_p[i] * scratch
+                                                   .phi_p[j]))
+                                               * scratch.finite_element_values.JxW(
+                                                 q);
+            }
+        }
+
+
+
+        void
+        local_assemble_newton_system_compressible (const double                                     pressure_scaling,
+                                                   const bool                                       rebuild_stokes_matrix,
+                                                   internal::Assembly::Scratch::StokesSystem<dim>  &scratch,
+                                                   internal::Assembly::CopyData::StokesSystem<dim> &data) const
+        {
+          const Introspection<dim> &introspection = this->introspection();
+          const unsigned int dofs_per_cell = scratch.finite_element_values.get_fe().dofs_per_cell;
+          const unsigned int n_q_points    = scratch.finite_element_values.n_quadrature_points;
+
+          for (unsigned int q=0; q<n_q_points; ++q)
+            {
+              for (unsigned int k=0; k<dofs_per_cell; ++k)
+                {
+                  scratch.phi_u[k] = scratch.finite_element_values[introspection.extractors.velocities].value (k,q);
+                  scratch.phi_p[k] = scratch.finite_element_values[introspection.extractors.pressure].value (k, q);
+                  if (rebuild_stokes_matrix)
+                    {
+                      scratch.grads_phi_u[k] = scratch.finite_element_values[introspection.extractors.velocities].symmetric_gradient(k,q);
+                      scratch.div_phi_u[k]   = scratch.finite_element_values[introspection.extractors.velocities].divergence (k, q);
+                    }
+                }
+
+              // Viscosity scalar
+              const double eta = (rebuild_stokes_matrix
```

Comment:
If you assembler the Newton system, you will *always* have to rebuild the Stokes matrix. So here and below, simplify the code by assuming that the variable is true.